app/src/main/cpp/ControllerManager.cpp

```diff
@@ -0,0 +1,479 @@
+}
+
+
+TogglePtr
+ControllerManager::GetRoot() const {
+  return m.root;
+}
+
+
+void
+ControllerManager::LoadControllerModel(const int32_t aModelIndex, const ModelLoaderAndroidPtr& aLoader, const std::string& aFileName) {
+  m.SetUpModelsGroup(aModelIndex);
+  aLoader->LoadModel(aFileName, m.models[aModelIndex]);
+}
+
+void
+ControllerManager::InitializePointer() {
+  if (m.pointerModel) {
+    return;
+  }
+  CreationContextPtr create = m.context.lock();
+  VertexArrayPtr array = VertexArray::Create(create);
+  const float kLength = -5.0f;
+  const float kHeight = 0.0008f;
+
+  array->AppendVertex(Vector(-kHeight, -kHeight, 0.0f)); // Bottom left
+  array->AppendVertex(Vector(kHeight, -kHeight, 0.0f)); // Bottom right
+  array->AppendVertex(Vector(kHeight, kHeight, 0.0f)); // Top right
+  array->AppendVertex(Vector(-kHeight, kHeight, 0.0f)); // Top left
+  array->AppendVertex(Vector(0.0f, 0.0f, kLength)); // Tip
+
+  array->AppendNormal(Vector(-1.0f, -1.0f, 0.0f).Normalize()); // Bottom left
+  array->AppendNormal(Vector(1.0f, -1.0f, 0.0f).Normalize()); // Bottom right
+  array->AppendNormal(Vector(1.0f, 1.0f, 0.0f).Normalize()); // Top right
+  array->AppendNormal(Vector(-1.0f, 1.0f, 0.0f).Normalize()); // Top left
+  array->AppendNormal(Vector(0.0f, 0.0f, -1.0f).Normalize()); // in to the screen
+
+
+  RenderStatePtr state = RenderState::Create(create);
+  state->SetMaterial(Color(0.6f, 0.0f, 0.0f), Color(1.0f, 0.0f, 0.0f), Color(0.5f, 0.5f, 0.5f),
+                     96.078431f);
+  GeometryPtr geometry = Geometry::Create(create);
+  geometry->SetVertexArray(array);
+  geometry->SetRenderState(state);
+
+  std::vector<int> index;
+  std::vector<int> uvIndex;
+
+  index.push_back(1);
+  index.push_back(2);
+  index.push_back(5);
+  geometry->AddFace(index, uvIndex, index);
+
+  index.clear();
+  index.push_back(2);
+  index.push_back(3);
+  index.push_back(5);
+  geometry->AddFace(index, uvIndex, index);
+
+  index.clear();
+  index.push_back(3);
+  index.push_back(4);
+  index.push_back(5);
+  geometry->AddFace(index, uvIndex, index);
+
+  index.clear();
+  index.push_back(4);
+  index.push_back(1);
+  index.push_back(5);
+  geometry->AddFace(index, uvIndex, index);
+
+  m.pointerModel = std::move(geometry);
+  for (Controller& controller: m.list) {
+    if (controller.transform) {
+      controller.transform->AddNode(m.pointerModel);
+    }
+  }
+}
+
+void
+ControllerManager::SetGestureDelegate(const GestureDelegateConstPtr& aGestureDelegate) {
+  m.gestures = aGestureDelegate;
+}
+
+
+void
+ControllerManager::Reset() {
+  for (Controller& controller: m.list) {
+    if (controller.transform) {
+      controller.transform->RemoveFromParents();
+    }
+    controller.Reset();
+  }
+  m.gestures = nullptr;
+}
+
+void
+ControllerManager::UpdateControllers(const float aFarClip, const std::vector<WidgetPtr>& aWidgets, const TrayPtr& aTray, bool& aRelayoutWidgets) {
```

Comment:
I'm wondering if this function should remain part of BrowserWorld? It seem like a lot of state to pass in only to use the controller's transformMatrix and buttons?. It seems like controller manager shouldn't need to know about widgets, or trays, or relayout. BrowserWorld feels like the better place for all these relatively independent pieces to meet.

Reply:
I'm ok with leaving it in BrowserWorld to reduce the dependencies. I'll update the PR.